data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -162,104 +165,238 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+                handleFailure(e, currentBufferPerBatch, resultRecords);
+                currentBufferPerBatch = lambdaCommonHandler.createBuffer(currentBufferPerBatch);
             }
         }
     }
 
-    protected Event convertLambdaResponseToEvent(InvokeResponse lambdaResponse) {
+    /*
+     * 1. If response has an array, we assume that we split the individual events.
+     * 2. If it is not an array, then create one event per response.
+     */
+    void convertLambdaResponseToEvent(
+            final List<Record<Event>> resultRecords,
+            final InvokeResponse lambdaResponse, Buffer flushedBuffer) {
         try {
+            // Current Assumption: Lambda always returns json array.
+            // TODO: add support for jsonObject/string/object return from lambda
+            boolean isJsonCodec = responseCodec instanceof JsonInputCodec;
+
+            //Get payload and convert to response codec
             SdkBytes payload = lambdaResponse.payload();
-            if (payload != null) {
-                String payloadJsonString = payload.asString(StandardCharsets.UTF_8);
-
-                JsonNode jsonNode = null;
-                try {
-                    jsonNode = objectMapper.readTree(payloadJsonString);
-                } catch (JsonParseException e) {
-                    throw new RuntimeException("payload output is not json formatted");
+
+            // Handle null or empty payload
+            if (payload == null || payload.asByteArray() == null || payload.asByteArray().length == 0) {
+                LOG.error(NOISY, "Lambda response payload is null or empty");
+                throw new RuntimeException("Lambda response payload is null or empty");
+            }
+
+            // Record payload sizes
+            requestPayloadMetric.set(flushedBuffer.getPayloadRequestSize());
+            responsePayloadMetric.set(payload.asByteArray().length);
+
+            LOG.debug("Response payload:{}",payload.asUtf8String());
+            InputStream inputStream = new ByteArrayInputStream(payload.asByteArray());
+            List<Event> parsedEvents = new ArrayList<>();
+            responseCodec.parse(inputStream, record -> {
+                Event event = record.getData();
+                parsedEvents.add(event);
+            });
+            if(parsedEvents.isEmpty()){
+                LOG.error("Parsing response failed for event, response:{}", payload.asUtf8String());
+                throw new RuntimeException("Response parsing failed");
+            }
+
+            List<Record<Event>> originalRecords = flushedBuffer.getRecords();
+
+            LOG.debug("Parsed Event Size:{}, FlushedBuffer eventCount:{}, FlushedBuffer size:{}",parsedEvents.size(),flushedBuffer.getEventCount(),flushedBuffer.getSize());
+            // Check if the response is a JSON array and the codec is JSON
+            if (isJsonCodec) {
+                if (parsedEvents.size() == flushedBuffer.getEventCount()) {
```

Comment:
We should use the strategy pattern to inject code for handling the parsed events based on aggregate or strict. Then you don't need this large conditional.

Something like this:

```
responseStrategy.handleEvents(parsedEvents, originalRecords);
```

Also, this will grow with some upcoming changes to support aggregations in end-to-end acknowledgements.